colossus-tests/src/test/scala/colossus/protocols/http/HttpPipeSpec.scala

```diff
@@ -0,0 +1,91 @@
+package colossus
+package protocols.http
+
+import core._
+import controller.Pipe
+import controller.PushResult
+
+import org.scalatest._
+
+import akka.util.ByteString
+
+import service.Callback
+import scala.util.{Try, Success, Failure}
+
+
+
+class FoldTester[T](foldCB: Callback[T]) {
+  private var res: Option[Try[T]] = None
+  foldCB.execute{result => res = Some(result)}
+
+  private def expectRes(r: Try[T] => Unit) { 
+    res.map(r).getOrElse{
+      throw new Exception("Callback did not complete")
+    }
+  }
+
+  def expect(expected: T) {
+    expectRes{
+      case Success(v) => if (v != expected) throw new Exception(s"Callback expected $expected, got $v")
+      case Failure(err) => throw new Exception(s"Expected $expected, but callback failed with $err")
+    }
+  }
+
+  def expectFailure() {
+    expectRes{
+      case Success(v) => throw new Exception("Expected Callback failure, but it completed with $v")
+      case Failure(_) => {}
+    }
+  }
+}
+
+class HttpPipeSuite extends WordSpec with MustMatchers{
+    
+  val chunkedBody = ByteString("3\r\nfoo\r\nb\r\n12345678901\r\n0\r\n\r\n")
+  val badChunkedBody = ByteString("3\r\nfoo\r\n1WHATTT\r\n12345678901\r\n0\r\n\r\n")
+
+  def byteFolder(buffer: DataBuffer, build: ByteString) = build ++ ByteString(buffer.takeAll)
+
+  def foldTester[A](p: Pipe[A, DataBuffer]): FoldTester[ByteString] = new FoldTester(p.fold(ByteString())(byteFolder))
+
+  "ChunkPassthroughPipe" must {
+
+    "leave the data untouched" in {
+      val p = new ChunkPassThroughPipe
+      val c = foldTester(p)
+      p.push(DataBuffer(chunkedBody)) must equal(PushResult.Complete)
+      c.expect(chunkedBody)
+    }
+
+    "properly detect the end of the stream" in {
+      val p = new ChunkPassThroughPipe
+      val c = foldTester(p)
+      p.push(DataBuffer(chunkedBody ++ ByteString("ADSFASDFASDF"))) must equal(PushResult.Complete)
```

Comment:
Is there anyway to assert that the ADSF string is till in the buffer, and has not been dropped?  I know you were concerned about that case.